### PR TITLE
feat(core): working implementation of indexable anchors Phase 2

### PR DESCRIPTION
**Pull request by Jovonni**
Working implementation of phase 2 of the indexable anchors work

- Added version to stream.ts
- changed snapshots to match new structure
- added @ethersproject/abi dependency
- added ABI, and interface to ethereum-anchor-validator
- added map for block_threshholds map
- added validation result type
- abstracted getting transaction from cache or RPC in ethereum-anchor-validator
- added conditional check for block number

### CODE REVIEW

File: packages/stream-model-handler/src/__tests__/model-handler.test.ts [hunk 1]
```diff
@@ -993,4 +1050,132 @@ describe('ModelHandler', () => {
...
+  //   //   commit: { proof: FAKE_CID_3, prev: FAKE_CID_2 },
+  //   //   proof: anchorProof,
+  //   // }
+  //   // state = await handler.applyCommit(anchorCommitData, context, state)
+  //   // delete state.metadata.unique
+  //   // expect(state).toMatchSnapshot()
+  // })
+
+  //could merge into anchor 0 test
+  it('applies version 1 anchor commit correctly', async () => {
```
Comment (stephhuynh18) [inline packages/stream-model-handler/src/__tests__/model-handler.test.ts:1114]: Sorry i'm a bit confused. I don't believe `handler.applyCommit` validates the commit. Please let me know if I am mistaken, or if you are testing something else.

**Review by stephhuynh18 — COMMENTED** [review]
As this is a draft please disregard any comments you are already addressing

File: packages/stream-model-handler/src/__tests__/model-handler.test.ts (hunk 1 above)
Comment (stbrody) [reply packages/stream-model-handler/src/__tests__/model-handler.test.ts:1114]: yeah I agree these tests aren't really adding a lot of value.  I think a more interesting test would be to update the InMemoryAnchorService to mirror the behavior of the contract-based anchoring system, then the existing tests in the `stream-tests` package that actually anchor real streams against the in-memory anchor service would test the new contract codepaths.  That would be a larger change though, and is still not as valuable as the actual e2e integration tests that use a real anchor service, which we'll be getting anyway and will actually be an easier change.  So I'd maybe file a ticket for updating the InMemoryAnchorService to use smart contract anchoring, but consider it a stretch goal for the project and not really worry about it right now.

The real valuable test coverage we are going to get is from the https://github.com/ceramicnetwork/ceramic-anchor-service/blob/develop/src/__tests__/ceramic_integration.test.ts test in the CAS codebase using Ganache, and from the e2e tests from https://github.com/3box/ceramic-integration-tests that use a real CAS against an Eth testnet

Comment (stbrody) [reply packages/stream-model-handler/src/__tests__/model-handler.test.ts:1114]: > I can knock this out

Let's focus on getting the core changes to js-ceramic and the CAS codebase in first, then we can go back and shore up test coverage second.  The sooner we can get this running in the QA environment the better, then we can add more test coverage as it burns in on QA

Comment (stephhuynh18) [reply packages/stream-model-handler/src/__tests__/model-handler.test.ts:1114]: We could add a new test file for the ethereum validator service because i don't see any tests that are not mocking the validation step. But I agree with spencer, lets focus on getting the core changes because there are dependencies

File: package.json [hunk 2]
```diff
@@ -1,42 +1,131 @@
...
+		"eslint-config-3box": "^0.4.0",
+		"eslint-plugin-jest": "^25.7.0",
+		"jest": "^27.4.7",
+		"jest-resolver-enhanced": "^1.0.1",
+		"lerna": "^4.0.0",
+		"prettier": "^2.5.1",
+		"rimraf": "^3.0.2",
+		"typedoc": "^0.22.11",
+		"typescript": "^4.5.4"
+	},
```
Comment (ukstv) [inline package.json:23]: That's not right. `package.json` should not be changed this much. You sure you are using NPM v8 + Node v16?

File: packages/cli/src/ceramic-cli-utils.ts [hunk 3]
```diff
@@ -36,7 +36,7 @@ const DEFAULT_DAEMON_CONFIG = DaemonConfig.fromObject({
   ipfs: { mode: IpfsMode.BUNDLED },
   logger: { 'log-level': LogLevel.important, 'log-to-files': false },
   metrics: { 'metrics-exporter-enabled': DEFAULT_METRICS_EXPORTER_ENABLED, 'metrics-port': DEFAULT_METRICS_EXPORTER_PORT },
-  network: { name: Networks.TESTNET_CLAY },
+  network: { name: Networks.LOCAL },
```
Comment (ukstv) [inline packages/cli/src/ceramic-cli-utils.ts:39]: Why?

File: packages/common/src/stream.ts [hunk 4]
```diff
@@ -62,7 +62,10 @@ export interface AnchorProof {
   blockNumber: number
   blockTimestamp: number
   txHash: CID
+  // txHash: CID | string
```
Comment (ukstv) [inline packages/common/src/stream.ts:65]: dev artefact. should not be here.

File: packages/common/src/stream.ts [hunk 5]
```diff
@@ -62,7 +62,10 @@ export interface AnchorProof {
...
+  // txHash: string
   root: CID
+  version?: number
```
Comment (ukstv) [inline packages/common/src/stream.ts:68]: Nice.

File: packages/core/package.json [hunk 6]
```diff
@@ -55,6 +55,7 @@
     "@stablelib/sha256": "^1.0.1",
     "await-semaphore": "^0.1.3",
     "dids": "^3.2.0",
+    "ganache": "^7.3.2",
```
Comment (ukstv) [inline packages/core/package.json:58]: `ganache` as a prod dependency? o_O It should be a dev dependency.

File: packages/core/src/anchor/ethereum/CeramicAnchorServiceV2.json [hunk 7]
```diff
@@ -0,0 +1,1746 @@
+{
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/CeramicAnchorServiceV2.json:1]: suggestion: rename that to `CeramicAnchorServiceV2.abi.json`.

File: packages/core/src/anchor/ethereum/CeramicAnchorServiceV2.json [hunk 8]
```diff
@@ -0,0 +1,1746 @@
...
+          "type": "address"
+        }
+      ],
+      "name": "transferOwnership",
+      "outputs": [],
+      "stateMutability": "nonpayable",
+      "type": "function"
+    }
+  ],
+  "bytecode": {
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/CeramicAnchorServiceV2.json:164]: Also, is it necessary to put here whole artefact, rather than just ABI?

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 9]
```diff
@@ -6,6 +6,9 @@ import { AnchorProof, AnchorValidator, DiagnosticsLogger } from '@ceramicnetwork
 import { Block, TransactionResponse } from '@ethersproject/providers'
 import { base16 } from 'multiformats/bases/base16'
 
+import { Interface } from '@ethersproject/abi';
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:9]: Should be included to package.json

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 10]
```diff
@@ -22,9 +25,17 @@ interface EthNetwork {
...
   'eip155:1': { network: 'mainnet', chain: 'ETH', chainId: 1, networkId: 1, type: 'Production' },
-  'eip155:3': { network: 'ropsten', chain: 'ETH', chainId: 3, networkId: 3, type: 'Test' },
+  
+  //PROD
+  // 'eip155:3': { network: 'ropsten', chain: 'ETH', chainId: 3, networkId: 3, type: 'Test' },
+  
   'eip155:4': { network: 'rinkeby', chain: 'ETH', chainId: 4, networkId: 4, type: 'Test' },
   'eip155:5': { network: 'goerli', chain: 'ETH', chainId: 5, networkId: 5, type: 'Test' },
+  
+  //DEV
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:35]: dev artefact again

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 11]
```diff
@@ -93,10 +110,20 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
+      console.log("this._getEthProvider(chainId)")
+      console.log(this._getEthProvider(chainId))
       const provider: providers.BaseProvider = this._getEthProvider(chainId)
+      console.log("provider")
+      console.log(provider)
       let transaction = this._transactionCache.get(txHash)
+      console.log("transaction x")
+      console.log(transaction)
 
       if (!transaction) {
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:123]: This thing is better moved to a separate function.

**Review by ukstv — CHANGES_REQUESTED** [review]

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 12]
```diff
@@ -137,32 +161,64 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
    */
-  async validateChainInclusion(anchorProof: AnchorProof): Promise<void> {
+  async validateLegacy(anchorProof: AnchorProof): Promise<ValidationResult> {
     const decoded = decode(anchorProof.txHash.multihash.bytes)
     const txHash = '0x' + uint8arrays.toString(decoded.digest, 'base16')
-
     const [transaction, block] = await this._getTransactionAndBlockInfo(anchorProof.chainId, txHash)
     const txValueHexNumber = parseInt(transaction.data, 16)
     const rootValueHexNumber = parseInt('0x' + anchorProof.root.toString(base16), 16)
+    return new this.ValidationResult(transaction, block, txValueHexNumber, rootValueHexNumber)
```
Comment (stephhuynh18) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:173]:
```suggestion
    return { txResponse: transaction, block, txValueHexNumber, rootValueHexNumber }

```
```suggestion
    return new this.ValidationResult(transaction, block, txValueHexNumber, rootValueHexNumber)
```

Comment (stephhuynh18) [reply packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:173]: no need for this or new

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 13]
```diff
@@ -137,32 +161,64 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
-    if (anchorProof.blockTimestamp !== block.timestamp) {
+    if (anchorProof.blockTimestamp !== validationResult.block.timestamp) {
       throw new Error(
-        `Block timestamps are not the same. AnchorProof blockTimestamp: ${anchorProof.blockTimestamp}, eth txn blockTimestamp: ${block.timestamp}`
+        `Block timestamps are not the same. AnchorProof blockTimestamp: ${anchorProof.blockTimestamp}, eth txn blockTimestamp: ${validationResult.block.timestamp}`
       )
     }
+
+    const BLOCK_THRESHHOLD = 1 //1000000000; //put into config, check where
+    if (validationResult.txResponse.blockNumber < BLOCK_THRESHHOLD) {
```
Comment (stephhuynh18) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:218]: This checks if the block number is less than the threshold. This should be checking if the block  number is greater than the threshold and the version is 0 or non existent.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts [hunk 14]
```diff
@@ -323,6 +323,8 @@ export class InMemoryAnchorService implements AnchorService, AnchorValidator {
       blockTimestamp: timestamp,
       txHash: CID.parse(SAMPLE_ETH_TX_HASH),
       root: leaf.cid,
+      // @note: force version 1 here? should it matter if its in memory
```
[the remark this answers is not in the document]
Comment (stephhuynh18) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:326]: Also file a ticket

File: packages/stream-model-handler/src/__tests__/__snapshots__/model-handler.test.ts.snap [hunk 15]
```diff
@@ -191,6 +191,197 @@ Object {
 }
 `;
 
+exports[`ModelHandler applies anchor version 1 commit correctly 1`] = `
```
Comment (stephhuynh18) [inline packages/stream-model-handler/src/__tests__/__snapshots__/model-handler.test.ts.snap:194]: This snapshot should be removed as there are no associated tests anymore

**Review by stephhuynh18 — CHANGES_REQUESTED** [review]
A suggestion for a later cleanup PR: `validateLegacy` and `validateContract` are technically not validating, but rather parsing the transaction data in different ways. So to make things clearer we can rename thing to 'parseXXX` and move things around accordingly

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts (hunk 11 above)
Comment (ukstv) [reply packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:124]: `if (!transaction) {...` that puts the transaction in the cache should also be there in `_getTransaction`.

The goal is to hide the caching.

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 16]
```diff
@@ -137,32 +162,66 @@ export class EthereumAnchorValidator implements AnchorValidator {
   }
 
   /**
-   * Validate anchor proof on the chain
+   * Validate version 0 anchor proof on the chain by the reading tx data directly
    * @param anchorProof - Anchor proof instance
    */
-  async validateChainInclusion(anchorProof: AnchorProof): Promise<void> {
+  async parseAnchorProofLegacy(anchorProof: AnchorProof): Promise<ValidationResult> {
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:168]: It is v0. Legacy is too ambiguous.

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 17]
```diff
@@ -137,32 +162,66 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
     const txValueHexNumber = parseInt(transaction.data, 16)
     const rootValueHexNumber = parseInt('0x' + anchorProof.root.toString(base16), 16)
+    return { txResponse: transaction, block, txValueHexNumber, rootValueHexNumber }
+  }
+
+  /**
+   * Validate version 1 anchor proof on the chain by parsing first encoded parameter
+   * @param anchorProof - Anchor proof instance
+   */
+  async parseAnchorProofV2(anchorProof: AnchorProof): Promise<ValidationResult> {
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:181]: And this is v1.

File: packages/core/src/ceramic.ts [hunk 18]
```diff
@@ -70,8 +70,8 @@ const SUPPORTED_CHAINS_BY_NETWORK = {
   [Networks.MAINNET]: ['eip155:1'], // Ethereum mainnet
   [Networks.ELP]: ['eip155:1'], // Ethereum mainnet
   [Networks.TESTNET_CLAY]: ['eip155:3', 'eip155:4', 'eip155:100'], // Ethereum Ropsten, Rinkeby, Gnosis Chain
-  [Networks.DEV_UNSTABLE]: ['eip155:3', 'eip155:4', 'eip155:5'], // Ethereum Ropsten, Rinkeby, Goerli
-  [Networks.LOCAL]: ['eip155:1337'], // Ganache
+  [Networks.DEV_UNSTABLE]: ['eip155:3', 'eip155:4', 'eip155:5'], // Ethereum Ropsten, Rinkeby, Goerli  
+  [Networks.LOCAL]: ['eip155:3', 'eip155:1337'], // Ganache
```
Comment (ukstv) [inline packages/core/src/ceramic.ts:74]: Why is it `eip155:3`?

File: package.json [hunk 19]
```diff
@@ -1,42 +1,42 @@
...
-    "typedoc": "^0.22.11",
-    "typescript": "^4.5.4"
-  },
-  "prettier": "eslint-config-3box/prettier.config",
-  "dependencies": {
-    "node-fetch": "npm:@achingbrain/node-fetch@^2.6.6"
-  }
-}
+	"name": "root",
+	"scripts": {
```
Comment (ukstv) [inline package.json:3]: tabs vs spaces.

Spaces were here before. Please, return it back to spaces.

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 20]
```diff
@@ -137,32 +162,66 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
+    const decodedArgs = iface.decodeFunctionData('anchor', transaction.data)
+    const rootCID = decodedArgs[0]
+    const txValueHexNumber = parseInt(rootCID, 16)
+    const rootValueHexNumber = parseInt('0x' + anchorProof.root.toString(base16), 16)  
+    return { txResponse: transaction, block, txValueHexNumber, rootValueHexNumber }
+  }
 
-    if (txValueHexNumber !== rootValueHexNumber) {
+  async validate(anchorProof: AnchorProof): Promise<ValidationResult> {
+    if(anchorProof.version === 1){
```
Comment (ukstv) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:193]: Nitpick: prettier this all.

**Review by ukstv — CHANGES_REQUESTED** [review]

**Review by stephhuynh18 — COMMENTED** [review]
just noticed that small thing, and could add the TODO format we discussed that includes the tickets, also just gotta get these tests/linting to pass

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 21]
```diff
@@ -25,6 +26,7 @@ const ETH_CHAIN_ID_MAPPINGS: Record<string, EthNetwork> = {
   'eip155:3': { network: 'ropsten', chain: 'ETH', chainId: 3, networkId: 3, type: 'Test' },
   'eip155:4': { network: 'rinkeby', chain: 'ETH', chainId: 4, networkId: 4, type: 'Test' },
   'eip155:5': { network: 'goerli', chain: 'ETH', chainId: 5, networkId: 5, type: 'Test' },
+  'eip155:1337': { network: 'local', chain: 'ETH', chainId: 1337, networkId: 1337, type: 'Test' },
```
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:29]: is this still necessary?

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 22]
```diff
@@ -33,6 +35,28 @@ const MAX_PROVIDERS_COUNT = 100
 const TRANSACTION_CACHE_SIZE = 50
 const BLOCK_CACHE_SIZE = 50
 
+const ABI = ['function anchor(bytes)']
+
+const iface = new Interface(ABI)
+
+//TODO finalize block number
```
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:42]: add a linear issue and reference the ticket number where this will be fixed, ie `// TODO(NET-XXXX): Finalize block number once CAS is creating smart contract anchors`

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 23]
```diff
@@ -76,6 +100,22 @@ export class EthereumAnchorValidator implements AnchorValidator {
     this._chainId = chainId
   }
 
+  /**
+   * isoldated method for fetching tx from cache, and if not set
+   **/
+  private async _getTransaction(
+    provider: providers.BaseProvider,
+    chainId: string,
```
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:108]: is chainId used?  Can it be removed as an argument?

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 24]
```diff
@@ -137,32 +173,70 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
+    const [transaction, block] = await this._getTransactionAndBlockInfo(anchorProof.chainId, txHash)
+    const decodedArgs = iface.decodeFunctionData('anchor', transaction.data)
+    const rootCID = decodedArgs[0]
+    const txValueHexNumber = parseInt(rootCID, 16)
+    const rootValueHexNumber = parseInt('0x' + anchorProof.root.toString(base16), 16)
+    return { txResponse: transaction, block, txValueHexNumber, rootValueHexNumber }
+  }
 
-    if (txValueHexNumber !== rootValueHexNumber) {
+  async validate(anchorProof: AnchorProof): Promise<ValidationResult> {
```
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:203]: this isn't really validating anything, it's just parsing the information from the transaction.  Can we call it `parseTransactionData` or similar?

Comment (stbrody) [reply packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:203]: or I guess `parseAnchorProof` for symmetry with its internal helpers

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 25]
```diff
@@ -137,32 +173,70 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
+    if (
+      validationResult.txResponse.blockNumber > BLOCK_THRESHHOLDS[this._chainId] &&
+      (anchorProof.version === 0 || !anchorProof.version)
+    ) {
+      throw new Error(
+        `Any anchor proofs created after block ${BLOCK_THRESHHOLD} must include the version field. AnchorProof blockNumber: ${anchorProof.blockNumber}`
+      )
+    }
+
+    //TODO: check contract addresses matches
```
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:239]: add reference to linear ticket number

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts [hunk 26]
```diff
@@ -323,6 +323,8 @@ export class InMemoryAnchorService implements AnchorService, AnchorValidator {
       blockTimestamp: timestamp,
       txHash: CID.parse(SAMPLE_ETH_TX_HASH),
       root: leaf.cid,
+      //TODO: update the InMemoryAnchorService to mirror the behavior of the contract-based anchoring system
```
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:326]: linear ticket number in all TODOs

File: packages/core/src/ceramic.ts [hunk 27]
```diff
@@ -70,7 +70,7 @@ const SUPPORTED_CHAINS_BY_NETWORK = {
   [Networks.MAINNET]: ['eip155:1'], // Ethereum mainnet
   [Networks.ELP]: ['eip155:1'], // Ethereum mainnet
   [Networks.TESTNET_CLAY]: ['eip155:3', 'eip155:4', 'eip155:100'], // Ethereum Ropsten, Rinkeby, Gnosis Chain
-  [Networks.DEV_UNSTABLE]: ['eip155:3', 'eip155:4', 'eip155:5'], // Ethereum Ropsten, Rinkeby, Goerli
+  [Networks.DEV_UNSTABLE]: ['eip155:3', 'eip155:4', 'eip155:5'], // Ethereum Ropsten, Rinkeby, Goerli
```
Comment (stbrody) [inline packages/core/src/ceramic.ts:73]: can you revert the change to this file to keep the git history clean?

**Review by stbrody — COMMENTED** [review]
looking pretty good, a few small comments inline but getting close

**Review by ukstv — APPROVED** [review]
🏇

File: packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts [hunk 28]
```diff
@@ -137,32 +171,69 @@ export class EthereumAnchorValidator implements AnchorValidator {
...
+    // if the block number is greater than the threshold and the version is 0 or non existent
+    if (
+      validationResult.txResponse.blockNumber > BLOCK_THRESHHOLDS[this._chainId] &&
+      (anchorProof.version === 0 || !anchorProof.version)
+    ) {
+      throw new Error(
+        `Any anchor proofs created after block ${BLOCK_THRESHHOLDS[this._chainId]} must include the version field. AnchorProof blockNumber: ${anchorProof.blockNumber}`
+      )
+    }
+    //TODO (NET-1657): Add check to validateAnchorInclusion for ensuring contract addresses match the official contract address
```
Comment (oed) [inline packages/core/src/anchor/ethereum/ethereum-anchor-validator.ts:236]: 👍

**Review by oed — APPROVED** [review]
lgtm

**Review by stbrody — APPROVED** [review]
LGTM.  Make sure to update the commit message before merging to match semver convention.  Should start with `feat(core): `

**Review by stephhuynh18 — APPROVED** [review]

**Comment by ukstv** [comment]
Yahoo!!
